text/0026-activation-distribution.md

```diff
@@ -0,0 +1,120 @@
+website in the browser.
+
+An **app shell** refers to the index.html page that includes the required base
+scripts. Thus triggering the whole loading process necessary for all the
+microfrontends.
+
+The **initial script** is the script performing the loading of the Single SPA
+library. It sets up important globals and imports the important global modules.
+
+The new **root config** replaces the old concept of a [root config][1]. An
+optional element of the import map, named `@openmrs/esm-root-config`, which can
+be used to (1) provide module configs to `@openmrs/esm-module-config`, (2)
+override module activator functions, and (3) do any other pre-app-launch setup
+wanted by the actual implemention. Removed responsibility: Registration of
+microfrontends with the Single SPA framework.
+
+A **microfrontend interface** describes the conditions on the exports of an ESM
+which, if met, indicate that the module is an OpenMRS microfrontend module. In
+particular, the modules exporting a `setupOpenMRS` are considered a
+microfrontend module.
+
+The interface for the `setupOpenMRS` function can be described in TypeScript as
+
+```ts
+interface SetupOpenMRS {
+  (): {
+	  lifecycle: LifeCycles<T> | (() => Promise<LifeCycles<T> | Splat<LifeCycles<T>>>);
+    activate(location: Location): boolean;
+  };
+}
+```
+
+The evaluation of the exports of the modules from the importmaps can be done in
+the initial script of the app shell.
+
+The process to migrate from a centralized module containing all activity
+functions to a distributed introduces a new `index.ts` in the microfrontends.
+
+The `index.ts` may be as simple as follows:
+
+```js
+// set the public path to be able to retrieve side bundles
+import "./set-public-path";
+// get useful helpers for the activity functions from the loaded root-config
+import { routePrefix } from '@openmrs/esm-root-config';
+
+export function setupOpenMRS() {
+  return {
+    lifecycle: () => import('./openmrs-esm-home'),
+    activate: location => routePrefix("home", location),
+  };
+}
+
+// export other dependencies to be used, e.g., by the dev tooling
+export { backendDependencies } from "./openmrs-backend-dependencies";
+```
+
+## Reason for decision
+- Enhance possibility of distributing the microfrontends
+- Make it possible to create self-containing microfrontends
+- Allow an easy packaging model for OpenMRS distributions
+- Use existing standards instead of creating new ones
+- Cooperate with existing architecture
+- Improve developer experience
+
+## Alternatives
+There are three main alternatives that can be seen:
+
+1. Switching away from Single SPA to microfrontend frameworks that support the
+notion of distributed microfrontends.
+2. Keeping the original notion of a single file that has the ultimate knowledge
+(with all the drawbacks and complexity for distribution).
+3. Using a service to get the information via an API call. Besides the original
```

Comment:
As of single-spa@5.3.0, single-spa accepts strings as activity functions. For example:

```js
singleSpa.registerApplication({
  name: 'nav',
  app: () => System.import('nav'),
  activeWhen: ['/']
})

singleSpa.registerApplication({
  name: 'patient-chart',
  app: () => System.import('patient-chart'),
  activeWhen: ['/patient/:uuid/chart']
})
```

- [Release notes](https://github.com/single-spa/single-spa/releases/tag/v5.3.0)
- [Documentation](https://single-spa.js.org/docs/configuration#configactivewhen)

^ I think that storing a root config in a service/database is a very viable approach, and is one that I know several organizations are already employing. The database could store the name of the service and a list of activeWhen strings - that is all that is needed to register the application.

Reply:
> If I understand this proposal correctly, I think that it would require downloading the code for all microfrontends in order to register the single-spa applications?

No - the microfrontends are bundle splitted as stated; so you only download essentially 0.5 KB - not the whole MF. So I assume this was misunderstood.

> As of single-spa@5.3.0, single-spa accepts strings as activity functions.

That's great and certainly useful also for the JSON configuration. However, it does not help with the distribution.

> I think I lean toward Alternative 3

Wonderful - thanks for the suggestion! :)

Reply:
> No - the microfrontends are bundle splitted as stated; so you only download essentially 0.5 KB - not the whole MF.

I've read through the proposal several times and do not see where it discusses code splitting. Could you help me understand this? Even with code splitting, this still comes with a performance cost - if there are 25 microfrontends, this proposal requires downloading the entry bundle for all 25 of them in order to register the applications?

Reply:
> That's great and certainly useful also for the JSON configuration. However, it does not help with the distribution.

Could you help me understand this? Distributions already have a variety of configuration stored in the database that is accessible via the Admin tool. This includes which Java modules are activated or not, which is fairly similar to which javascript modules are activated. I agree that storing code in a database has downsides, but don't see how it prevents distribution, since my understanding of creating OpenMRS distributions is that this often involves setting up a database with correct config.

Reply:
> if there are 25 microfrontends, this proposal requires downloading the entry bundle for all 25 of them in order to register the applications?

Right now we have multiple loose scripts via importmaps, which could also be served as a single UMD. You cannot argue in favor of fixed importmaps (+ include 6 loose SystemJS scripts, all which could be served as a single UMD, too), but not be in favor of microfrontends being loosely coupled. Microfrontends are all about loose coupling.

Furthermore, this only needs to happen once. The script can then be cached.

I am not sure if I read all that correctly - to me it reads with a certain attitude (maybe I am mistaking here). I propose to discuss this in a call. I guess it would make more sense to you then.

Reply:
My intent here is engaging in a respectful, productive dialog. I am not arguing with you, but seeking to understand. Thanks for the [clarification comment](https://github.com/openmrs/openmrs-rfc-frontend/pull/26#discussion_r420865829) below calling out the code split via `import()`, that helps me understand how this proposal works 👍.

> You cannot argue in favor of fixed importmaps (+ include 6 loose SystemJS scripts, all which could be served as a single UMD, too), but not be in favor of microfrontends being loosely coupled. Microfrontends are all about loose coupling.

I don't view putting top level routes into a root config as tight coupling. And do not see the connection between loading SystemJS scripts and the current discussion here.

#### Summary of my concern

The current proposal requires that every microfrontend's entry bundle be downloaded in order to register microfrontends. The entry bundle can be quite small due to code splits, but it still is a fairly large and dynamic number of network requests before the first meaningful render. I think that a fairly large, dynamic number of network requests has a performance penalty that should be avoided if possible.

#### Maybe another alternative?

Perhaps instead of downloading all modules in the browser in order to retrieve their activity functions, each package could publish to npm a separate file that could be included into the root-config as a build-time dependency?

Reply:
Let me try to show you the connection to SystemJS (which is just one example). You refer about saving requests, which obviously contradict the current design (e.g., instead of a single bundle we serve 7 scripts, 6 references, 1 inline). I would love to set the same metrics everywhere.

> each package could publish to npm a separate file that could be included into the root-config as a build-time dependency?

Here I think there is a misunderstanding. You have several independently developed and deployed modules. How can they now come together in a build (coming from a system that is shipped pre-built and does not have any knowledge of which and how many modules are actually used later on)?

The idea of this proposal is to allow a free composition.

Reply:
> You refer about saving requests, which obviously contradict the current design (e.g., instead of a single bundle we serve 7 scripts, 6 references, 1 inline). I would love to set the same metrics everywhere.

The difference between the systemjs scripts and this proposal is that the systemjs scripts are a fixed 5-6 requests, whereas the number of applications is dynamic. The systemjs library itself is quite small (~5kb gzipped for everything).

A dynamic number of requests has a lot more potential to impact performance than a fixed number of requests.

> Here I think there is a misunderstanding. You have several independently developed and deployed modules. How can they now come together in a build (coming from a system that is shipped pre-built and does not have any knowledge of which and how many modules are actually used later on)?

Most OpenMRS distributions do not independently deploy modules, but rather package them all up and ship them out as `.war` files. In order for deployments to be possible, there must be centralization of servers, which is not true of the majority of the OpenMRS community. As far as I know, there are only very few organizations (Ampath being one of them) that have centralized their servers.

I agree that my proposal disallows for independent deployment, which is not ideal. That's a valid criticism and perhaps the idea should be scrapped 😄.

My intent here is to try to find a way to avoid an unknown and dynamic number of network requests before the first meaningful render.

Reply:
> My intent here is to try to find a way to avoid an unknown and dynamic number of network requests before the first meaningful render.

I think this is good and obviously the argument with static vs dynamic number is also true. Nevertheless, the argument that these initial requests should be already bundled is in the same sense also right - and justifies that we are not playing with the same rules here. Unfortunately.

Nevertheless, coming back to the topic. The thinking with the unknown number of requests does not take into consideration HTTP/2 batching and multiplexing. The cost of multiple requests for scripts that are ~less than 1 KB in size is in reality quite negligible. And again - that these requests exist is only true for FTU. Afterwards, they are all there - which by the way also improves update scenarios. Instead of two things you'll only need to update one module. In this case only the updated single script is requested from the browser.

Reply:
Fair enough - if it is ensured that the code splitting occurs, http2 enabled, gzip, etc, perhaps the performance impact is worth the benefits of splitting this out.

I'll leave this conversation here for now and let the MF core team evaluate this PR further. I am not as involved in openmrs as I was last year and view my role here as just providing discussion points / insight, but not as the decision maker.

I think you're a smart person @FlorianRappl who clearly has some experience and knows your stuff. I'm glad OpenMRS has you to help push things forward.